pkg/maps/nat/stats/stats.go

```diff
@@ -0,0 +1,320 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package stats
+
+import (
+	"container/heap"
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/cilium/cilium/pkg/controller"
+	"github.com/cilium/cilium/pkg/ipcache"
+	"github.com/cilium/cilium/pkg/maps/nat"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/promise"
+	"github.com/cilium/cilium/pkg/statedb"
+	"github.com/cilium/cilium/pkg/tuple"
+	"github.com/cilium/cilium/pkg/u8proto"
+
+	"github.com/cilium/hive/cell"
+	"github.com/cilium/hive/job"
+
+	"github.com/sirupsen/logrus"
+)
+
+type Stats struct {
+	metrics natMetrics
+
+	db    *statedb.DB
+	table statedb.RWTable[NatMapStats]
+
+	logger   logrus.Entry
+	maxPorts int
+	config   Config
+	natMap4  nat.NatMap4
+	natMap6  nat.NatMap6
+	ipcache  *ipcache.IPCache
+}
+
+type NatMapStats struct {
+	Type       string
+	EgressIP   string
+	EndpointIP string
+	RemotePort uint16
+	Proto      string
+	Count      int
+	Nth        int
+}
+
+func (s NatMapStats) addrs() (string, string) {
+	if s.Type == nat.IPv6.String() {
+		return "[" + s.EgressIP + "]", "[" + s.EndpointIP + "]"
+	}
+	return s.EgressIP, s.EndpointIP
+}
+
+func (NatMapStats) TableHeader() []string {
+	return []string{"IPFamily", "Proto", "EgressIP", "RemoteAddr", "Count"}
+}
+
+func (s NatMapStats) TableRow() []string {
+	var raddr string
+	eip, rip := s.addrs()
+	if s.RemotePort == 0 {
+		raddr = rip
+	} else {
+		raddr = fmt.Sprintf("%s:%d", rip, s.RemotePort)
+	}
+	return []string{s.Type, s.Proto, eip, raddr, strconv.Itoa(s.Count)}
+}
+
+type params struct {
+	cell.In
+
+	Lifecycle cell.Lifecycle
+	DB        *statedb.DB
+	Table     statedb.RWTable[NatMapStats]
+	NatMap4   promise.Promise[nat.NatMap4]
+	NatMap6   promise.Promise[nat.NatMap6]
+	Jobs      job.Group
+	Metrics   natMetrics
+	Config    Config
+	Health    cell.Health
+}
+
+func newStats(params params) (*Stats, error) {
+	if params.Config.NATMapStatInterval == 0 {
+		return nil, nil
+	}
+
+	if params.Config.NatMapStatKStoredEntries > maxNatMapStatKStoredEntries ||
+		params.Config.NatMapStatKStoredEntries < minNatMapStatKStoredEntries {
+		return nil, fmt.Errorf("nat-stats config: %q must be betweed [%d, %d]",
+			natMapStatsEntriesName, minNatMapStatKStoredEntries, maxNatMapStatKStoredEntries)
+	}
+
+	maxAvailPorts := 65535 - (option.Config.NodePortMax + 1)
```

Comment:
Does this mean `NODEPORT_PORT_MAX_NAT` - `NODEPORT_PORT_MIN_NAT`?
https://github.com/cilium/cilium/blob/1c9d968f832659452977cede8b86db68a6697bb4/pkg/datapath/linux/config/config.go#L546

Can we define a constraint for NODEPORT_PORT_MAX_NAT somewhere and comment that it corresponds to  `NODEPORT_PORT_MAX_NAT` - `NODEPORT_PORT_MIN_NAT`?